src/fromager/resolver.py

```diff
@@ -220,50 +297,23 @@ def find_matches(
         identifier_reqs = list(requirements[identifier])
         bad_versions = {c.version for c in incompatibilities[identifier]}
 
+        candidates = self.get_from_cache(identifier, bad_versions, identifier_reqs)
+        if candidates:
+            return sorted(
+                candidates, key=attrgetter("version", "build_tag"), reverse=True
+            )
+
         # Need to pass the extras to the search, so they
         # are added to the candidate at creation - we
         # treat candidates as immutable once created.
-        candidates = []
         for candidate in get_project_from_pypi(
             identifier, set(), self.sdist_server_url
         ):
-            # Skip versions that are known bad
-            if candidate.version in bad_versions:
-                if DEBUG_RESOLVER:
-                    logger.debug(
-                        f"{identifier}: skipping bad version {candidate.version} from {bad_versions}"
-                    )
-                continue
-            # Skip versions that do not match the requirement
-            if not all(candidate.version in r.specifier for r in identifier_reqs):
-                if DEBUG_RESOLVER:
-                    logger.debug(
-                        f"{identifier}: skipping {candidate.version} because it does not match {identifier_reqs}"
-                    )
-                continue
-            # Skip versions that do not match the constraint
-            if not self.constraints.is_satisfied_by(identifier, candidate.version):
-                if DEBUG_RESOLVER:
-                    c = self.constraints.get_constraint(identifier)
-                    logger.debug(
-                        f"{identifier}: skipping {candidate.version} due to constraint {c}"
-                    )
-                continue
-            # Only include sdists if we're asked to
-            if candidate.is_sdist and not self.include_sdists:
-                if DEBUG_RESOLVER:
-                    logger.debug(
-                        f"{identifier}: skipping {candidate} because it is an sdist"
-                    )
-                continue
-            # Only include wheels if we're asked to
-            if not candidate.is_sdist and not self.include_wheels:
-                if DEBUG_RESOLVER:
-                    logger.debug(
-                        f"{identifier}: skipping {candidate} because it is a wheel"
-                    )
-                continue
-            candidates.append(candidate)
+            if not self.is_in_cache(identifier, candidate) and self.is_valid_candidate(
+                candidate, identifier, bad_versions, identifier_reqs
+            ):
+                candidates.append(candidate)
+                self.add_to_cache(identifier, candidate)
```

Comment:
How would we get a candidate that's already in the cache here?

Can we hide that check by having `add_to_cache()` perform it? That would make it work more like a `set()`.

This also feels like it is going to cache some candidates, but not all, and not the actual resolved result. 

We could build a new class that takes as input a resolver instance and provides the same interface for looking up versions that the resolver does, but that saves the results before returning them. That would be a relatively small class, I think?

Reply:
We get a candidate that is in cache here: https://github.com/python-wheel-build/fromager/pull/358/files#diff-7dc38361431a0d4763f2cae35a298b8b64a0885f49485d4911096dca9d6e9a4bR300

And right now we will are only storing candidates in cache that were actually potential candidates for some package condition. Do you want me store all seen candidates in the cache?

Reply:
I was thinking we would only cache the resolved _results_, not the candidates.

Reply:
I can do that

Reply:
I just realized that we initialize a new provider object each time we do resolution. So we will need to define a global cache and pass that to the provider or initialize the provider only once

Reply:
That feels like another reason to keep the cache on the outside of the provider.

We have an [override method](https://github.com/python-wheel-build/fromager/blob/main/src/fromager/sources.py#L101) for getting per-package resolver providers, so we can't have a singleton resolver.

However, we would expect those unique providers to give us the same results, and a cache hit could mean we avoid invoking any provider at all.

We're only looking at using the cache for bootstrap, right? Because for the build we are given an exact version and there's no need to resolve it. So maybe we should look for a place where we could put a cache that would only show up in the code path used for bootstrapping?

Reply:
We use it here as well: https://github.com/python-wheel-build/fromager/blob/main/src/fromager/commands/build.py#L175

Reply:
Nvm caching won't make a difference here so yeah just bootstrap

Reply:
I think we will still need to access the cache inside the provider because only the provider knows how to check if a candidate is can be used or not

Reply:
Why do we need to re-evaluate each of the previously seen candidates?

Reply:
You're thinking of the evaluation criteria other than the version?

Reply:
yep all the different if checks we do apart from just checking if the version matches the requirement